### ISSUE TITLE
image_picker error:Permission Denial

### ISSUE THREAD

**Issue by StarOfLife** [flutter/flutter#29422]
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, v1.2.1, on Microsoft Windows [Version 10.0.17763.379], locale zh-CN)
[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[!] Android Studio (version 3.3)
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
[√] VS Code (version 1.32.3)
[√] Connected device (1 available)

PlatformException (PlatformException(error, Permission Denial: starting Intent { act=android.media.action.IMAGE_CAPTURE flg=0x3 cmp=com.android.camera2/com.android.camera.CaptureActivity clip={text/uri-list U:content://com.example.st.flutter.image_provider/external_files/Android/data/com.example.st/files/Pictures/b213cf0a-aabb-45a5-9f36-bbc61deac3547272842760043634351.jpg} (has extras) } from ProcessRecord{fc5f102 8235:com.example.sht/u0a79} (pid=8235, uid=10079) with revoked permission android.permission.CAMERA, null))

 image_picker: 0.5.0+6
image_picker is remove unnecessary camera permmision, if I have‘t allow app to use camera before open,it will be crash without show PermissionDialog

**Comment by worldofsites** [comment]
Same issue.
Reproduced on image_picker: 0.5.0+6
Not reproduced on image_picker: 0.5.0+3

Reproduced until the application does not access the camera.

```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel master, v1.3.13-pre.10, on Microsoft Windows [Version 10.0.17763.379], locale ru-RU)
[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[√] Android Studio (version 3.3)
[√] VS Code (version 1.32.2)
[√] Connected device (1 available)

• No issues found!
```

**Comment by iampawan** [comment]
@pepegich @mklim Why you guys removed the camera permission? Now, how someone is supposed to take the picture from the camera without asking for permission?

**Comment by pepegich** [comment]
@iampawan I run example from plugin on nexus 5 android 4.4.4 adn pixel 2 android 9. Crash did not occur.
check comment: https://github.com/flutter/flutter/issues/19806#issuecomment-408343729

**Comment by arok** [comment]
The issue reproduces only if you specify `android.permission.CAMERA` in your AndroidManifest.xml or if another library (for example barcode_scaner) specify this permission.

**Comment by StarOfLife** [comment]
@pepegich the example is ok,and i find the error. in another library it specify android.permission.CAMERA. thank you

**Comment by StarOfLife** [comment]
@arok .thank you

**Comment by worldofsites** [comment]
@StarOfLife 
Why are you close that ticket?
How you resolve that issue?

If I have application which depends on barcode_scan and image_picker packages. And barcode_scan require android.permission.CAMERA in AndroidManifest.xml, but image_picker doesn't require that permission.  What I should do?
@pepegich @mklim

**Comment by StarOfLife** [comment]
@worldofsites I'm so sorry, I don't think about use both barcode_scan and image_picker

**Comment by arok** [comment]
@worldofsites you need manually check/request camera permission, see this package https://pub.dartlang.org/packages/permission_handler

**Comment by iampawan** [comment]
I solved the above issue by actually implementing the permission manually because I was also using Barcode Scan package. However, I just want to know @pepegich that having that permission for these kinds of use-cases can fix this problem. What do you say? Will it make sense? You would be knowing better 😀

**Comment by worldofsites** [comment]
@arok Thanks!

**Comment by wreppun** [comment]
I appreciate that there is a workaround, but this seems a little odd to me.

If you _do not_ have `<uses-permission android:name="android.permission.CAMERA" />` in `AndroidManifest.xml`, image_picker works.

If you _do_ have `<uses-permission android:name="android.permission.CAMERA" />`, then you either need to:

a. Request permission manually before calling ImagePicker.pickImage the first time
b. Use another plugin which requires camera permissions first, and then use the ImagePicker.

But if you

c. Try to use ImagePicker.pickImage without a or b, it will crash flutter.

-----------

It seems to me that changing the configuration/installation of the plugin (adding a permission request), depending on if a _different_ plugin is using the camera, would be a little confusing.

**Comment by von-maurus** [comment]
> I appreciate that there is a workaround, but this seems a little odd to me.
> 
> If you _do not_ have `<uses-permission android:name="android.permission.CAMERA" />` in `AndroidManifest.xml`, image_picker works.
> 
> If you _do_ have `<uses-permission android:name="android.permission.CAMERA" />`, then you either need to:
> 
> a. Request permission manually before calling ImagePicker.pickImage the first time
> b. Use another plugin which requires camera permissions first, and then use the ImagePicker.
> 
> But if you
> 
> c. Try to use ImagePicker.pickImage without a or b, it will crash flutter.
> 
> It seems to me that changing the configuration/installation of the plugin (adding a permission request), depending on if a _different_ plugin is using the camera, would be a little confusing.

Yes, it worked for me. the upgrade busted up my project XD!

**Comment by maral** [comment]
Unfortunately, this same error is happening to me even when I don't have the camera permission stated in `AndroidManifest.xml`, neither do I use camera anywhere else. 

Currently I'm using version 0.5.2, Android API 28.

What might be going on?

My error:
java.lang.SecurityException: Permission Denial: starting Intent { act=android.media.action.IMAGE_CAPTURE flg=0x3 cmp=com.android.camera2/com.android.camera.CaptureActivity clip={text/uri-list U:content://my.package.name.flutter.image_provider/external_files/Android/data/my.package.name/files/Pictures/7171ae3d-ae2c-40c0-bfeb-d1aa2a613cf05402625956103201743.jpg} (has extras) } from ProcessRecord{9ff798f 9542:cz.codit.stromysvobody/u0a85} (pid=9542, uid=10085) with revoked permission android.permission.CAMERA

**Comment by maral** [comment]
Even manually asking for permission did not help me (although I'm not sure I did it correctly). Downgrading to 0.5.0+3 solved the issue.

**Comment by von-maurus** [comment]
> Unfortunately, this same error is happening to me even when I don't have the camera permission stated in `AndroidManifest.xml`, neither do I use camera anywhere else.
> 
> Currently I'm using version 0.5.2, Android API 28.
> 
> What might be going on?
> 
> My error:
> `java.lang.SecurityException: Permission Denial: starting Intent { act=android.media.action.IMAGE_CAPTURE flg=0x3 cmp=com.android.camera2/com.android.camera.CaptureActivity clip={text/uri-list U:content://my.package.name.flutter.image_provider/external_files/Android/data/my.package.name/files/Pictures/7171ae3d-ae2c-40c0-bfeb-d1aa2a613cf05402625956103201743.jpg} (has extras) } from ProcessRecord{9ff798f 9542:cz.codit.stromysvobody/u0a85} (pid=9542, uid=10085) with revoked permission android.permission.CAMERA`

When you erased the permission stated in AndroidManifest.xml, also did you delete the 
"build" generated folder? When you change the manifest, you have to also delete that folder, reset android studio or any IDE and build the project again.

Other possibility can be that the permission implementation that you did its wrong.

**Comment by maral** [comment]
I actually never had the permission in AndroidManifest.xml at all and got the error. Only after checking upon this issue I tried adding it there and it has not fixed it nor change the behaviour at all.

**Comment by simeonzv** [comment]
Thanks to @arok, there is a workaround using https://pub.dartlang.org/packages/permission_handler

And here is an example on how it can be implemented

```dart
import 'package:permission_handler/permission_handler.dart';

Future<bool> checkAndRequestCameraPermissions() async {
  PermissionStatus permission =
      await PermissionHandler().checkPermissionStatus(PermissionGroup.camera);
  if (permission != PermissionStatus.granted) {
    Map<PermissionGroup, PermissionStatus> permissions =
        await PermissionHandler().requestPermissions([PermissionGroup.camera]);
    return permissions[PermissionGroup.camera] == PermissionStatus.granted;
  } else {
    return true;
  }
}
```
Then every call to the image picker has to be wrapped like this to prevent your app from crashing

```dart
if (await checkAndRequestCameraPermissions()) {
  File image = await ImagePicker.pickImage(source: ImageSource.camera);
  // continue with the image ...
}
```
But I still prefer that something like this is done within the image_picker plugin.

**Comment by pepegich** [comment]
I fixed, https://github.com/flutter/plugins/pull/1519

**Comment by cyanglaz** [comment]
Should be fixed by https://github.com/flutter/plugins/pull/1519. Closing this issue now. If any related issue has been found after applying using the newest Image_picker, please open a another issue with details.

**Comment by DjivaCanessane** [comment]
Hello, 
I am trying to publish my app, which is using SMS: ^0.2.4, on the Play Store.
In my androidManifest.xml file I have declared that my app requires three permissions:

```
<uses-permission android:name="android.permission.CAMERA" />
    <uses-permission android:name="android.permission.RECORD_AUDIO" />
    <uses-permission android:name="android.permission.READ_CONTACTS" />
```

When I upload it on the play console, it tells me that I have to declare 3 other sensitive permission:

android.permission.READ_SMS
android.permission.SEND_SMS
android.permission.RECEIVE_SMS

I think it's due to the SMS plugin, and I need it to invite other users to join my app.
I receive this mail from Google: 
<img width="836" alt="Capture d’écran 2019-05-12 à 02 09 02" src="https://user-images.githubusercontent.com/50156885/57576241-0cf8f900-745b-11e9-82dc-b7567719b577.png">
Could someone help me with this issue please ?

**Comment by cyanglaz** [comment]
@DjivaCanessane maybe try to post an issue on the SMS plug-in github page?

**Comment by tapizquent** [comment]
I am getting this same error in iOS. I was testing and denied the permissions, and I have yet to find a way to request permission again if they do an action.

`[VERBOSE-2:ui_dart_state.cc(148)] Unhandled Exception: PlatformException(photo_access_denied, The user did not allow photo access., null)`

I am doing this in an iOS simulator. Do I really have to delete the app and reinstall to allow permissions?

**Comment by kaiserleka** [comment]
Same issue like @tapizquent  on Android simulator and real device

`var image = await ImagePicker.pickImage(source: ImageSource.gallery);`

`Exception has occurred.
PlatformException (PlatformException(photo_access_denied, The user did not allow photo access., null))`

dependencies:
```
flutter:
    sdk: flutter
  firebase_auth: ^0.14.0+5
  cloud_firestore: ^0.12.9+4
  http: ^0.12.0+2
  package_info: ^0.4.0+6
  image_picker: ^0.6.1+4
```
Not worked even if I locate this or not 
`<uses-permission android:name="android.permission.CAMERA" />`

**Comment by kaiserleka** [comment]
I seems these permissions solved my problem.

`<uses-permission android:name="android.permission.MANAGE_DOCUMENTS" />`
`<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />`

The point I haven't understand. I had been accessing device gallery without any permission before.

**Comment by lukepighetti** [comment]
@kaiserleka Are you requesting permissions for those with something like `permission_handler` explicitly? I can't get `image_picker` to work on Android 10.

**Comment by p02rosea** [comment]
Hi! I have found another error when picking image from gallery or photo on API 29, it works fine on API 28 but on 29 it can not access to the file picture.

`Unable to decode stream: java.io.FileNotFoundException: /storage/emulated/0/Download/descarga.jpeg: open failed: EACCES (Permission denied)`

**Comment by lifenautjoe** [comment]
Same issue here. Do we really this permission ?

```
<uses-permission android:name="android.permission.MANAGE_DOCUMENTS" />
```

**Comment by everystone** [comment]
I just ran into this, I got it working by adding 
```
    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
    <uses-permission android:name="android.permission.MANAGE_DOCUMENTS" />
```

then deleting the build folder and rebuilding the app.

**Comment by srihamat** [comment]
@everystone Many thanks, keyword delete build folder is really helped me.

**Comment by aseem-hegshetye** [comment]
I was pointed to this page multiple times but I havent found any solution yet. 
So I will state my specifications:
I have `image_picker: ^0.6.2+2`  `Flutter 1.9.1+hotfix.6` . I am trying to do `var image = await ImagePicker.pickImage(source: ImageSource.gallery);`  I get error `photo_access_denied, The user did not allow photo access` . How do I allow photo access in android emulator?

**Comment by sandeepgurram** [comment]
I was getting issue because of build folder. When I run "flutter clean", issue was fixed.

**Comment by xudre** [comment]
I fixed this error for **>= Android Q** adding `android:requestLegacyExternalStorage="true"` to the `<application>` in **AndroidManifest.xml**.

**Comment by lunedam-git** [comment]
This is actually still an ongoing problem as far as I can tell.

Using 0.6.3+4, my app requires me to use requestLegacyExternalStorage=true - even for the 'Google' Photo and Download directories in order to successfully use ImageSource.gallery as my source - on both a simulator and a Samsung S10+, both running Android 10. I thought that access was allowed to the Photo and Download directories under scoped storage? Apparently not though, as in my case (and many others it appears) Legacy External Storage is required even there. And with Android 'R' making legacy external storage requests will no longer be allowed at all, as I understand it, so the requestLegacyExternalStorage=true setting in AndroidManifest will be redundant too for future apps.

Confusing times indeed.

**Comment by ratherNotB** [comment]
I had the same problem. Turned out I hadn't followed the instructions below to upgrade my pre-Flutter v1.12 project:

https://github.com/flutter/flutter/wiki/Upgrading-pre-1.12-Android-projects

After upgrading my project my problems were resolved.

**Comment by mishatron** [comment]
`for >= Android Q adding android:requestLegacyExternalStorage="true" to the <application> in AndroidManifest.xml.` - authors, could you please add this information to official library documentation?

**Comment by jaween** [comment]
> I had the same problem. Turned out I hadn't followed the instructions below to upgrade my pre-Flutter v1.12 project:
> 
> https://github.com/flutter/flutter/wiki/Upgrading-pre-1.12-Android-projects
> 
> After upgrading my project my problems were resolved.

This fixed the issue for me too. I think this is the correct solution and the only solution for supporting Android 11.

**Comment by lunedam-git** [comment]
> I had the same problem. Turned out I hadn't followed the instructions below to upgrade my pre-Flutter v1.12 project:
> 
> https://github.com/flutter/flutter/wiki/Upgrading-pre-1.12-Android-projects

Unfortunately this did **not** fix the problem for me, using v0.6.4. I still have to include android:requestLegacyExternalStorage="true" to avoid a 'permission denied' error when trying to pick an image. So I guess my app will never work in Android 11 - or I must find an alternative image picker somewhere. Failing that I'll have to stop offering the gallery as a choice to Android users. Which is mighty disappointing to say the least.

**Comment by ratherNotB** [comment]
In that case I would suggest creating a new workspace and generating a
fresh project in it using 'flutter create'. Then you'd need to copy your
code over and apply any Android and iOS configurations that you've made...
such as package name, etc.

I had to resort to this once. It took a couple of hours but resolved my
issues.
